src/blueapi/data_management/gda_directory_provider.py

```diff
@@ -0,0 +1,145 @@
+        return DataCollectionIdentifier(collectionNumber=self._count)
+
+    async def get_current_collection(self) -> DataCollectionIdentifier:
+        return DataCollectionIdentifier(collectionNumber=self._count)
+
+
+class VisitDirectoryProvider(DirectoryProvider):
+    """
+    Gets information from a remote service to construct the path that detectors
+    should write to, and determine how their files should be named.
+    """
+
+    _data_group_name: str
+    _data_directory: Path
+
+    _client: VisitServiceClientBase
+    _current_collection: Optional[DirectoryInfo]
+    _session: Optional[ClientSession]
+
+    def __init__(
+        self,
+        data_group_name: str,
+        data_directory: Path,
+        client: VisitServiceClientBase,
+    ):
+        self._data_group_name = data_group_name
+        self._data_directory = data_directory
+        self._client = client
+
+        self._current_collection = None
+        self._session = None
+
+    async def update(self) -> None:
+        """
+        Calls the visit service to create a new data collection in the current visit.
+        """
+        # TODO: After visit service is more feature complete:
+        # TODO: Allow selecting visit as part of the request to BlueAPI
+        # TODO: Consume visit information from BlueAPI and pass down to this class
+        # TODO: Query visit service to get information about visit and data collection
+        # TODO: Use AuthN information as part of verification with visit service
+
+        try:
+            collection_id_info = await self._client.create_new_collection()
+            self._current_collection = self._generate_directory_info(collection_id_info)
+        except Exception as ex:
+            # TODO: The catch all is needed because the RunEngine will not
+            # currently handle it, see
+            # https://github.com/bluesky/bluesky/pull/1623
+            self._current_collection = None
+            logging.exception(ex)
+
+    def _generate_directory_info(
+        self,
+        collection_id_info: DataCollectionIdentifier,
+    ) -> DirectoryInfo:
+        collection_id = collection_id_info.collectionNumber
+        file_prefix = f"{self._data_group_name}-{collection_id}"
+        return DirectoryInfo(str(self._data_directory), file_prefix)
+
+    def __call__(self) -> DirectoryInfo:
+        if self._current_collection is not None:
+            return self._current_collection
+        else:
+            raise ValueError(
+                "No current collection, update() needs to be called at least once"
+            )
+
+
+_SINGLETON: Optional[VisitDirectoryProvider] = None
+
+
+def set_directory_provider_singleton(provider: VisitDirectoryProvider):
+    global _SINGLETON
+
+    _SINGLETON = provider
+
+
+def get_directory_provider() -> DirectoryProvider:
+    if _SINGLETON is not None:
+        return _SINGLETON
+    else:
+        return StaticDirectoryProvider(tempfile.TemporaryFile(), "")
```

Comment:
I can't seem to find any instances of this being used. It seems like when we start up blue api and a handler is made, in that big 'if' block that checks for config, we make a provider (if we have config) and then add attach_metadata to the plan wrappers with the provider we've just made. With that in mind, is there any point to making the directory provider in blue api a global variable?

This goes hand in hand with my earlier comment about set_directory_provider_singleton not actually being used.

so perhaps the best thing to do, is get rid of this block (from line 132 to 145) and don't fiddle with a global directory provider; just set the one in dodal.

Reply:
You're right, this should be removed. Will put a fuller explanation in https://github.com/DiamondLightSource/blueapi/pull/315#discussion_r1371443992